pydb/oracle_db.py

```diff
@@ -0,0 +1,182 @@
+#!/usr/bin/env python3
+"""
+Module for interacting with an oracle database
+"""
+# -*- encoding: utf-8 -*-
+
+#============================================================================================
+# Imports
+#============================================================================================
+# Standard imports
+import logging
+
+# Third-party imports
+import cx_Oracle
+
+from pylog.pylog import get_common_logger_for_module
+
+from .database import DBInterface
+
+# Local imports
+
+
+class OracleDB(DBInterface):
+    """
+    Class for interacting with an oracle database
+    """
+
+    def __init__(self, host: str, port: int, service: str, user: str, pwd: str,
+                 logging_level: int = 50, logging_format: logging.Formatter = None):
+        """
+        Setup for oracle db connections. oracle_config must be a python dictionary with the following fields:
+
+        :param host:
+        :param port:
+        :param service:
+        :param user:
+        :param pwd:
+        :param logging_level:
+        :param logging_format:
+        :param logging_level: defaults to logging.CRITICAL
+        :param logging_format: defaults to None here, which translates to the pylog.get_commong_logging_format
+        """
+
+        self.host = host
+        self.port = port
+        self.service = service
+        self.user = user
+        self.pwd = pwd
+        self.pool = self.set_up_session_pool()
+
+        self.logger = get_common_logger_for_module(module_name=__name__, level=logging_level, log_format=logging_format)
+
+    def set_up_session_pool(self):
+        try:
+            dsn_str = cx_Oracle.makedsn(self.host, self.port, service_name=self.service)
+            pool = cx_Oracle.SessionPool(
+                user=self.user,
+                password=self.pwd,
+                dsn=dsn_str,
+                min=2,
+                max=5,
+                increment=1,
+                threaded=True,
+                encoding="UTF-8"
+            )
+            self.pool = pool
+            return pool
+
+        except cx_Oracle.DatabaseError as err:
+            obj, = err.args
+            self.logger.error("Error creating pool")
+            self.logger.error("Context: %s", obj.context)
+            self.logger.error("Message: %s", obj.message)
+            raise Exception(f"Error creating pool: {obj.message}")
+
+    def get_session_pool(self):
+        """
+        Function for creating a session pool with the database
+        """
+        if self.pool:
+            return self.pool
+        else:
+            self.set_up_session_pool()
```

Comment:
The else clause here technically will never fire because you always set self.pool in `__init__` or throw an error.

One alternative would be to just use a property for this to load it on demand:

```
class OracleDB(DBInterface):
    def __init__(self, host: str, port: int, service: str, user: str, pwd: str,
                       logging_level: int = 50, logging_format: logging.Formatter = None):

        . . .

        self._pool = None

        . . .
        
    @property
    def pool(self):
        if not self._pool:
           self._pool = self.set_up_session_pool()

        return self._pool
```

Reply:
updated